Ozone Wayland IME integration

zhani opened the pull request:
Introduce Ozone Wayland IME integration. This patch
exposes Linux input method context service that can be
used with ime service on Linux platforms. Ozone Wayland
platform implements Linux input method context service
and integrates it with zwp_text_input_v1 protocol.

#155

jdapena commented on ui/ozone/platform/wayland/wayland_input_method_context.cc:
```diff
@@ -0,0 +1,103 @@
+}
+
+static base::string16 ConvertToUTF16(const std::string& text) {
+  if (base::IsStringUTF8(text)) {
```
Why this? You should assume always UTF-8 on strings coming from text input interface.

Literally from specs: "      Text is generally UTF-8 encoded, indices and lengths are in bytes." https://cgit.freedesktop.org/wayland/wayland-protocols/tree/unstable/text-input/text-input-unstable-v1.xml#n41

jdapena commented on chrome/browser/ui/views/ime_driver/input_method_bridge_linux.h:
```diff
@@ -0,0 +1,38 @@
+// This bridges between mojo InputMethod API and aura::InputMethodBridgeLinux.
+// It forwards the received events to an instance of
+// aura::InputMethodBridgeLinux.
+class InputMethodBridgeLinux : public ui::mojom::InputMethod {
```
Very likely this could be moved to ui/. But for that, RemoteTextInputClient also needs to be moved to ui/. Maybe to ui/base/ime?

I've been checking and actually all IMEDriver could be moved to ui.

zhani replied:
Agree but I don't think it's a good idea to do it in this commit

jdapena replied:
Agree. Also, very likely this refactor could be landed upstream earlier than the desktop wayland specific changes. So that refactor could be directly contributed upstream.

jdapena commented on ui/ozone/platform/wayland/wayland_input_method_context.cc:
```diff
@@ -0,0 +1,103 @@
+WaylandInputMethodContext::WaylandInputMethodContext(
+    WaylandConnection* connection)
+    : connection_(connection), text_input_(nullptr) {
+  use_ozone_wayland_vkb_ = getenv("USE_OZONE_WAYLAND_VKB");
```
Better using a command line switch. It should be possible to just add one in ozone_switches. Maybe a set of --enable-ozone-ime/--disable-ozone-ime, that would be enabled/disabled by default depending on the platform (i.e. I guess it should be disabled in all ozone platforms right now?). Not sure about ChromeOS default behavior in this case, as the IME they are exposing is not going through ozone right?

msisov commented on ui/ozone/platform/wayland/zwp_text_input_wrapper_v1.h:
```diff
@@ -0,0 +1,104 @@
+class ZWPTextInputWrapperV1 : public ZWPTextInputWrapper {
+ public:
+  ZWPTextInputWrapperV1(zwp_text_input_manager_v1* text_input_manager);
+  virtual ~ZWPTextInputWrapperV1();
```
nit: remove virtual, add override.

zhani replied:
Fixed

msisov reviewed (commented):
I stlill can't build it 

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: aura::LinuxInputMethodContextFactoryMus::LinuxInputMethodContextFactoryMus(service_manager::Connector*)
>>> referenced by memory:3026 (../../buildtools/third_party/libc++/trunk/include/memory:3026)
>>>               ime_driver_mus.o:(IMEDriver::Register()) in archive obj/chrome/browser/ui/libui.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.

msisov commented on ui/ozone/platform/wayland/BUILD.gn:
```diff
@@ -59,8 +64,10 @@ source_set("wayland") {
     "//third_party/wayland:wayland_client",
+    "//third_party/wayland-protocols:text_input_protocol",
     "//third_party/wayland-protocols:xdg_shell_protocol",
     "//ui/base",
```
add missing "//ui/events/keycodes:xkb",

zhani replied:
Fixed

msisov commented on services/ui/public/interfaces/ime/linux_input_method_context.mojom:
```diff
@@ -0,0 +1,52 @@
+  Initialize(LinuxInputMethodContextDelegate delegate, bool is_simple);
+
+  // Dispatches the key event to an underlying IME.  Returns true if the key
+  // event is handled, otherwise false.  A client must set the text input type
```
nit: s/false.  A/false. A/

tonikitoo commented on ui/base/ime/DEPS:
```diff
@@ -1,3 +1,3 @@
 include_rules = [
-  "+chromeos/system",
+  "+chromeos/system"
```
unneeded change?

tonikitoo commented on ui/ozone/platform/wayland/ozone_platform_wayland.cc:
```diff
@@ -124,6 +126,22 @@ class OzonePlatformWayland : public OzonePlatform {
+          const service_manager::BindSourceInfo&>* registry) override {
+    registry->AddInterface<ui::mojom::LinuxInputMethodContext>(
+        base::Bind(&OzonePlatformWayland::CreateInputMethodContext,
+                   weak_factory_.GetWeakPtr()));
```
can we used  base::Unretained(this) instead of a weak reference?

tonikitoo commented on ui/ozone/platform/wayland/wayland_connection.h:
```diff
@@ -39,9 +39,13 @@ class WaylandConnection : public PlatformEventSource,
 
   WaylandWindow* GetWindow(gfx::AcceleratedWidget widget);
   WaylandWindow* GetCurrentFocusedWindow();
+  WaylandWindow* GetCurrentKeyboardFocusWindow();
```
GetCurrentKeyboardFocusedWindow? (note: *Focused*)

tonikitoo commented on ui/ozone/platform/wayland/wayland_input_method_context_untitests.cc:
```diff
@@ -0,0 +1,148 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
this file contain a typo on its name: "xxx_untitests.cc".

tonikitoo commented on ui/ozone/platform/wayland/wayland_window.cc:
```diff
@@ -204,6 +204,8 @@ bool WaylandWindow::HasCapture() {
 }
 
 void WaylandWindow::Show() {
+  set_keyboard_focus(true);
```
does that affect tooltips? i.e. do we transfer focus keyboard to tooltips windows?

- for example, place focus on an input, part of chrome or content.
- place mouse cursor over an element with tooltip
(tooltip shows and keyboard focus is set to it).

does it have side effect? when is it keyboard unfocused?

tonikitoo commented on ui/ozone/platform/wayland/zwp_text_input_wrapper_v1.cc:
```diff
@@ -0,0 +1,187 @@
+void ZWPTextInputWrapperV1::OnEnter(void* data,
+                                    struct zwp_text_input_v1* text_input,
+                                    struct wl_surface* surface) {
+  // Not implemented
```
use NOTIMPLEMENTED()?

tonikitoo commented on ui/ozone/platform/wayland/zwp_text_input_wrapper_v1.cc:
```diff
@@ -0,0 +1,187 @@
+
+void ZWPTextInputWrapperV1::OnLeave(void* data,
+                                    struct zwp_text_input_v1* text_input) {
+  // Not implemented
```
use NOTIMPLEMENTED()?

tonikitoo commented on ui/ozone/platform/wayland/zwp_text_input_wrapper_v1.cc:
```diff
@@ -0,0 +1,187 @@
+    void* data,
+    struct zwp_text_input_v1* text_input,
+    uint32_t state) {
+  // Not implemented
```
use NOTIMPLEMENTED()?

tonikitoo commented on ui/ozone/public/ozone_switches.cc:
```diff
@@ -15,4 +15,8 @@ const char kOzoneDumpFile[] = "ozone-dump-file";
 const char kEnableDrmAtomic[] = "enable-drm-atomic";
 
+// Try to enable wayland input method editor.
+// This works only with wayland platform.
```
I think this second line could be omit. this first lines implicitly says it, and so does the command line name itself.

Commit:
third_party: Add text-input-unstable-v1 protocol to wayland-protocols.

This is a prerequisite to get Wayland IME for Ozone-Wayland working. Will be used
by an external window mode.

Bug: 791970, 791920
Change-Id: I1f2f09b8eafe5ff33d2ad6e1ca6f975ca4d6c051

tonikitoo commented on chrome/browser/ui/views/ime_driver/ime_driver_mus.cc:
```diff
@@ -16,12 +16,25 @@
 
 #if defined(OS_CHROMEOS)
 #include "chrome/browser/ui/views/ime_driver/input_method_bridge_chromeos.h"
+#elif defined(OS_LINUX) && defined(USE_OZONE) && !defined(OS_CHROMEOS)
```
strictly speaking, there is no need to check os_chromeos here, since it is already checked a couple of lines above?

tonikitoo commented on chrome/browser/ui/views/ime_driver/ime_driver_mus.cc:
```diff
@@ -50,6 +63,16 @@ void IMEDriver::StartSession(ui::mojom::StartSessionDetailsPtr details) {
   mojo::MakeStrongBinding(
       base::MakeUnique<InputMethodBridge>(std::move(remote_client)),
       std::move(details->input_method_request));
+#elif defined(OS_LINUX) && defined(USE_OZONE) && !defined(OS_CHROMEOS)
```
strictly speaking there is no need to check os_chromeos here, since it is already checked in line 56.

tonikitoo commented on services/ui/public/interfaces/ime/linux_input_method_context.mojom:
```diff
@@ -0,0 +1,52 @@
+import "services/ui/public/interfaces/ime/ime.mojom";
+
+// An interface of input method context for input method frameworks on
+// GNU/Linux and likes
```
nit: and likes, reads strange here.

tonikitoo commented on third_party/wayland-protocols/protocol/text-input-protocol-v1.c:
```diff
@@ -0,0 +1,92 @@
+/* Generated by wayland-scanner 1.12.0 */
```
nit: all other generated protocols seem to be using 1.13.0.

tonikitoo commented on ui/aura/mus/input_method_mus.cc:
```diff
@@ -179,6 +179,7 @@ void InputMethodMus::UpdateTextInputType() {
   ui::TextInputType type = GetTextInputType();
   ui::mojom::TextInputStatePtr state = ui::mojom::TextInputState::New();
   state->type = mojo::ConvertTo<ui::mojom::TextInputType>(type);
+
```
drop this change?

tonikitoo commented on ui/ozone/platform/wayland/wayland_connection.h:
```diff
@@ -55,6 +59,7 @@ class WaylandConnection : public PlatformEventSource,
                        const gfx::Point& location);
 
   int GetKeyboardModifiers();
+  void DispatchUiEvent(Event* event);
```
no need to make it public, if we pass it as a callback.

tonikitoo commented on ui/ozone/platform/wayland/wayland_input_method_context.cc:
```diff
@@ -0,0 +1,126 @@
+  bool down = state == WL_KEYBOARD_KEY_STATE_PRESSED;
+  ui::KeyEvent event(down ? ET_KEY_PRESSED : ET_KEY_RELEASED, key_code,
+                     dom_code, flags, dom_key, EventTimeForNow());
+  connection_->DispatchUiEvent(&event);
```
rather than calling WaylandConnection::DispatchUiEvent directly, can not we pass a Callback, like we do in WaylandKeyboard or WaylandPointer?


291       connection->keyboard_ = std::make_unique<WaylandKeyboard>(
292           keyboard, base::Bind(&WaylandConnection::DispatchUiEvent,
293                                base::Unretained(connection)));

zhani replied:
But in that case WaylandInputMethodContextTest and OzonePlatformWayland needs to be declared as friend classes for WaylandConnection, right? Is that better approach?

tonikitoo replied:
Hum, this is a good point. If other tests do this, it should be fine for us too.

tonikitoo commented on ui/ozone/platform/wayland/wayland_window.h:
```diff
@@ -56,6 +56,8 @@ class WaylandWindow : public PlatformWindow, public PlatformEventDispatcher {
     return has_pointer_focus_ || has_touch_focus_;
   }
 
+  bool has_keyboard_focus() { return has_keyboard_focus_; }
```
nit: more right below set_keyboard_focus declaration?

zhani replied:
Fixed

tonikitoo commented:
chromeos build fails with the patch:

````
$ time ninja -C out/ozone/chrome_os/ chrome  services_unittests ozone_unittests browser_tests mash:all 
ninja: Entering directory `out/ozone/chrome_os/'
[1/1] Regenerating ninja files
FAILED: build.ninja 
../../../home/tonikitoo/Devel/chromium/src/buildtools/linux64/gn --root=../../../home/tonikitoo/Devel/chromium/src -q gen .
ERROR Dependency cycle:
  //ui/ozone:ozone ->
  //ui/ozone:platform ->
  //ui/ozone/platform/wayland:wayland ->
  //services/ui/public/interfaces/ime:ime ->
  //ui/base/ime:ime ->
  //services/ui/public/cpp/input_devices:input_devices ->
  //ui/ozone:ozone

ninja: error: rebuilding 'build.ninja': subcommand failed

real	0m4,896s
user	0m9,630s
sys	0m2,442s

tonikitoo commented:
Also @zhani you have a pseudo green light on https://crrev.com/c/806376 (third_party: Add text-input-unstable-v1 protocol to wayland-protocols).

I suggest that we split this PR in at least two commits: one containing the wl protocol addition. another with the rest of the fix. It is easier for rebases purposes.

msisov commented:
bug: the virtual keyboard doesn't remove the existing characters if those has been typed before.

Commit:
Change conflicting mojo enum value

mus.mojom.KeyboardCode.KEY_DELETE conflicts
with KEY_DELETE macro in /usr/include/linux/input-event-codes.h.

https://codereview.chromium.org/1527183003 was supposed to
rename original mus.mojom.KeyboardCode.DELETE to DELETE_KEY but
it got renamed to KEY_DETELE which now conflicts.

This CL:
- mus.mojom.KeyboardCode.KEY_DELETE became DELETE_KEY

Commit:
IME for Mus: Linux IME context servification

Servification of Linux IME context allows implementing
IME context in Mus environment. This can be adopted to
existing LinuxInputMethodContext implementations and is
also an enabler for backends which need to run IME context
out of browser thread. Mainly targeted for Ozone Wayland
external mode that will run IME context in MUS thread.

This CL:
- Introduces LinuxInputMethodContext and
  LinuxInputMethodContextDelegate mojo service interfaces.
- Implements LinuxInputMethodContextMus and
  LinuxInputMethodContextFactoryMus.
- Groups ui/base/ime common types into their own GN component
  (ime_types) to break a dependency cycle on ChromeOS that
  this CL would cause otherwise.

Commit:
IME for Wayland: Ozone Wayland IME integration

Introduce Ozone Wayland IME integration.

This CL:
- Implements Ozone wayland specific Linux input method
  context service interface and integrates it with
  zwp_text_input_v1 protocol.
- Introduces new runtime flag --enable-wayland-ime to
  enable this feature (by default false).

Bug: 791920

Commit:
IME for Mus: Add ui::InputMethodAuraLinux logic to ime driver

Modify browser's IMEDriver so we have access to
ui::InputMethodAuraLinux functionality in mus

This CL:
- Modifies ui::InputMethodAuraLinux so it can call an ack callback after
  processing events.
- Introduces InputMethodBridge to acts as a bridge between
  ui::mojom::InputMethod and ui::InputMethodAuraLinux.
- Modifies content_browser's IMEDriver implementation to use
  ui::InputMethodAuraLinux via InputMethodBridge.
- Some unittests to verify that InputMethodBridge works correctly.

tonikitoo reviewed (commented):
commit message typo: must.mojom

tonikitoo commented on ui/aura/mus/linux_input_method_context_delegate_impl.h:
```diff
@@ -0,0 +1,48 @@
+class AURA_EXPORT LinuxInputMethodContextDelegateImpl
+    : public ui::mojom::LinuxInputMethodContextDelegate {
+ public:
+  LinuxInputMethodContextDelegateImpl(
```
please use 'explicit'

zhani replied:
fixed

tonikitoo commented on ui/aura/mus/linux_input_method_context_factory_mus.h:
```diff
@@ -0,0 +1,36 @@
+// found in the LICENSE file.
+
+#ifndef UI_AURA_MUS_LINUX_INPUT_METHOD_CONTEXT_FACTORY_MUS_H_
+#define UI_AURA_MUS_LINUX_INPUT_METHOD_CONTEXT_MUI_AURA_MUS_LINUX_INPUT_METHOD_CONTEXT_FACTORY_MUS_H_US_H_
```
the precompiler guard seems accidentally too long here. :)

zhani replied:
fixed

tonikitoo reviewed (approved):
lgtm

tonikitoo reviewed (commented):
maybe add to the commit message, how this .c/h files were generated (eg command lines)?

tonikitoo commented on ui/ozone/platform/wayland/wayland_input_method_context.h:
```diff
@@ -0,0 +1,54 @@
+class WaylandInputMethodContext : public ui::mojom::LinuxInputMethodContext,
+                                  public ZWPTextInputWrapperClient {
+ public:
+  WaylandInputMethodContext(WaylandConnection* connection);
```
explicit

zhani replied:
fixed

tonikitoo reviewed (commented):
lgtm. I believe we can upstream this one right away.

@zhani please  mention that compiler used, and it might be helpful to paste on the commit message
a condensed version of the compiler output.

tonikitoo commented on ui/ozone/platform/wayland/wayland_input_method_context.cc:
```diff
@@ -0,0 +1,126 @@
+
+namespace ui {
+
+const int kXkbKeycodeOffset = 8;
```
namespace ui {

namespace {

const int kXkbKeycodeOffset = 8;

}  // namespace

tonikitoo reviewed (commented):
lgtm

tonikitoo reviewed (commented):
lgtm

tonikitoo reviewed (commented):
lgtm

msisov commented on ui/aura/mus/linux_input_method_context_mus.cc:
```diff
@@ -0,0 +1,100 @@
+#include "ui/aura/mus/linux_input_method_context_delegate_impl.h"
+#include "ui/base/ime/text_input_client.h"
+
+#include "base/message_loop/message_loop.h"
```
move this upwards to other includes.

msisov commented on ui/aura/mus/linux_input_method_context_mus.h:
```diff
@@ -0,0 +1,70 @@
+  void DispatchKeyEventCallback(const ui::KeyEvent& event, bool handled);
+
+  // A set of callback functions.  Must not be nullptr.
+  std::unique_ptr<LinuxInputMethodContextDelegateImpl> delegate_;
```
nit: space after this line?

msisov commented on ui/ozone/platform/wayland/wayland_input_method_context.cc:
```diff
@@ -0,0 +1,130 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
2018?

msisov commented on ui/ozone/platform/wayland/wayland_input_method_context.cc:
```diff
@@ -0,0 +1,130 @@
+
+namespace {
+
+const int kXkbKeycodeOffset = 8;
```
constexpr

msisov commented:
it looks ok, but it fails to compile now.

msisov@debian-XPS:~/code/chromium/src$ ninja -C out/OzoneLinux/ chrome -j8
ninja: Entering directory `out/OzoneLinux/'
[21/174] SOLINK ./libui_base_ime.so
FAILED: libui_base_ime.so libui_base_ime.so.TOC 
python "../../build/toolchain/gcc_solink_wrapper.py" --readelf="readelf" --nm="nm" --sofile="./libui_base_ime.so" --tocfile="./libui_base_ime.so.TOC" --output="./libui_base_ime.so"  -- ../../third_party/llvm-build/Release+Asserts/bin/clang++ -shared -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -Wl,--no-as-needed -lpthread -Wl,--as-needed -fuse-ld=lld -Wl,--icf=all -m64 -Wl,-O2 -Wl,--gc-sections -Wl,--fatal-warnings -nostdlib++ --sysroot=../../build/linux/debian_stretch_amd64-sysroot -L../../build/linux/debian_stretch_amd64-sysroot/lib/x86_64-linux-gnu -Wl,-rpath-link=../../build/linux/debian_stretch_amd64-sysroot/lib/x86_64-linux-gnu -L../../build/linux/debian_stretch_amd64-sysroot/usr/lib/x86_64-linux-gnu -Wl,-rpath-link=../../build/linux/debian_stretch_amd64-sysroot/usr/lib/x86_64-linux-gnu -o "./libui_base_ime.so" -Wl,-soname="libui_base_ime.so" @"./libui_base_ime.so.rsp"
/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::InputMethodAuraLinux(ui::internal::InputMethodDelegate*))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::~InputMethodAuraLinux())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText(ui::CompositionText const&)
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::DispatchKeyEvent(ui::KeyEvent*, base::OnceCallback<void (bool)>))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::DispatchKeyEvent(ui::KeyEvent*, base::OnceCallback<void (bool)>))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::ProcessKeyEventDone(ui::KeyEvent*, base::OnceCallback<void (bool)>, bool, bool))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::ProcessKeyEventDone(ui::KeyEvent*, base::OnceCallback<void (bool)>, bool, bool))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::ResetContext())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::ResetContext())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::OnCommit(std::__1::basic_string<unsigned short, base::string16_internals::string16_char_traits, std::__1::allocator<unsigned short> > const&))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::OnCommit(std::__1::basic_string<unsigned short, base::string16_internals::string16_char_traits, std::__1::allocator<unsigned short> > const&))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::OnPreeditEnd())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::OnPreeditEnd())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::OnPreeditEnd())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::InputMethodAuraLinux::OnPreeditEnd())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::MockIMEInputContextHandler::MockIMEInputContextHandler())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::MockIMEInputContextHandler::~MockIMEInputContextHandler())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::CompositionText::~CompositionText()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(ui::MockIMEInputContextHandler::~MockIMEInputContextHandler())

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::ImeTextSpan::~ImeTextSpan()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(std::__1::enable_if<(__is_forward_iterator<ui::ImeTextSpan*>::value) && (is_constructible<ui::ImeTextSpan, std::__1::iterator_traits<ui::ImeTextSpan*>::reference>::value), void>::type std::__1::vector<ui::ImeTextSpan, std::__1::allocator<ui::ImeTextSpan> >::assign<ui::ImeTextSpan*>(ui::ImeTextSpan*, ui::ImeTextSpan*))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::ImeTextSpan::ImeTextSpan(ui::ImeTextSpan const&)
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(std::__1::enable_if<(__is_forward_iterator<ui::ImeTextSpan*>::value) && (is_constructible<ui::ImeTextSpan, std::__1::iterator_traits<ui::ImeTextSpan*>::reference>::value), void>::type std::__1::vector<ui::ImeTextSpan, std::__1::allocator<ui::ImeTextSpan> >::assign<ui::ImeTextSpan*>(ui::ImeTextSpan*, ui::ImeTextSpan*))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: undefined symbol: ui::ImeTextSpan::~ImeTextSpan()
>>> referenced by ime_jumbo_1.cc
>>>               obj/ui/base/ime/ime/ime_jumbo_1.o:(std::__1::enable_if<(__is_forward_iterator<ui::ImeTextSpan*>::value) && (is_constructible<ui::ImeTextSpan, std::__1::iterator_traits<ui::ImeTextSpan*>::reference>::value), void>::type std::__1::vector<ui::ImeTextSpan, std::__1::allocator<ui::ImeTextSpan> >::assign<ui::ImeTextSpan*>(ui::ImeTextSpan*, ui::ImeTextSpan*))

/home/msisov/code/chromium/src/out/OzoneLinux/../../third_party/llvm-build/Release+Asserts/bin/ld.lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[28/174] CXX obj/ui/ozone/platform/wayland/wayland/wayland_surface_factory.o
ninja: build stopped: subcommand failed.

msisov commented:
Good. Thank you for your contribution!
